### DISCUSSION TITLE
Multiple independent flows

### DISCUSSION THREAD

**Discussion by maxsteenbergen** [xyflow/xyflow#2068] [Q&A] [answered]
I have 2 flows in on page, between which I toggle using some state. Both instances load a different (but related) flow using `defaultNodes` and `defaultEdges`, which in turn are loaded via a `conf` prop and layouted with Elk.

For the first Flow, I load an array of items from an API and pass it to the `conf` prop when needed. 
Whenever the `conf` prop changes for this instance, React-Flow updates as expected.

For the second flow, the approach is much the same: 
 - I keep an index somewhere in a useState hook 
 - I updated the `conf` prop using this state (`conf={AvailableConfigs[HighlightedConfig].components}`)
 - I fetch the items and run it all through Elk.

It all fetches and layouts as you'd expect, but React-Flow just won't update the nodes entirely. It looks like it does update the `data` (labels and stuff change), but positions or even visibility all inherit from the first Flow instance.

I've tried: 
- Wrapping both Flows in their own Provider
- ensuring unique keys for each state
- separate components for both Flows
- useCallback vs useEffect
- internal state vs passed prop
- Make everything a new object every single pass
- lots of other stuff

but I just can't get the second flow to behave entirely independent from the first. I'm stumped.

The primary Flow:

```
 {Status !== null && (
     <LayoutFlow id="CurrentFlow" conf={Status} nodeTypes={nodeTypes}/>
)}
```

The secondary Flow component: 

```
<LayoutFlow
  id={"OptionalFlow" + AvailableConfigs[HighlightedConfig].id}  
  conf={AvailableConfigs[HighlightedConfig].components} 
  nodeTypes={nodeTypes}
/>
```

The actual Flow component
```
function LayoutFlow({id, conf, nodeTypes, style}) {
  const [Layout, setLayout] = useState({ nodes: [], edges: []})
  
  useEffect(()=>{
    const getLayout = async () => {
      const layout = await createGraphLayoutElk(conf, id)
      setLayout(await layout)
    }
    if(conf.length > 0)  getLayout()
  }, [conf, setLayout])
  
  
  return (
    <ReactFlowProvider key={id}>
      <ReactFlow
        defaultNodes={Layout.nodes}
        defaultEdges={Layout.edges}
        nodeTypes={nodeTypes}
        style={style}
      />
    </ReactFlowProvider>
  )
}
```

**Comment by moklick** [comment]
When you are working with `defaultNodes`/ `defaultEdges` you can't update them by altering them and then passing them again. It works like `defaultValue` of an `<input />` component. If you want to work with `defaultNodes`/ `defaultEdges`, you need to use the helper function of the [React Flow instance](https://reactflow.dev/docs/api/react-flow-instance) to update the nodes and edges.

**Reply by maxsteenbergen** [reply]
That's one of the approaches I tried: no difference.

To be complete: this code yields the same result (with the provider moved up a level)
```
function LayoutFlow({id, conf, index, nodeTypes, style}) {
  const FlowInstance = useReactFlow()
  
  useEffect(()=>{
    const getLayout = async () => {
      const components = conf[index].components ?? conf[index]
      const layout = await createGraphLayoutElk(components, id)
      FlowInstance.setNodes(layout.nodes)
      FlowInstance.setEdges(layout.edges)
    }
    if(conf.length > 0) getLayout()
  }, [conf, index])
  
  
  return (
    // <ReactFlowProvider>
      <ReactFlow
        defaultNodes={[]}
        defaultEdges={[]}
        nodeTypes={nodeTypes}
        style={style}
      />
    // </ReactFlowProvider>
  )
}
```

**Reply by moklick** [reply]
When you have two separate flows, you need two ReactFlowProviders in order to have a separate store for each of them.

**Reply by maxsteenbergen** [reply]
I tried that as well, at multiple places in the code: inside the component and outside of it. The result is the same

**Comment by maxsteenbergen** [comment, marked as answer]
Right, this can be resolved. Turns out I was misusing useMemo somewhere outside of React-Flow PLUS there was a subtle bug in the API we made. Sorry for the inconvenience!
